azkaban-web-server/src/main/java/azkaban/webapp/servlet/LoginAbstractAzkabanServlet.java

```diff
@@ -412,9 +413,11 @@ protected void handleAjaxLoginAction(final HttpServletRequest req,
       final Cookie cookie = new Cookie(SESSION_ID_NAME, session.getSessionId());
       cookie.setPath("/");
       resp.addCookie(cookie);
-      getApplication().getSessionCache().addSession(session);
+      SessionCache cache = getApplication().getSessionCache();
+      cache.addSession(session);
       ret.put("status", "success");
       ret.put("session.id", session.getSessionId());
+      ret.put("session.ttl", String.valueOf(cache.getEffectiveSessionTimeToLive()));
```

Comment:
This value is static per an AZ web server instance. Right?

Could you describe your use case in more details?

Reply:
Session TTL is a final/immutable value per AZ instance, it has a default value that can be overridden by a prop.
Use case for the TTL: once a client obtains a session id (token), it comes handy for the client to know how long the token is valid. E.g. not to to upload a large ZIP with an expired token (and be rejected, re-authenticate, and upload the ZIP again).
E.g. `linkedin-gradle-plugin-for-apache-hadoop` could benefit from having such information.

Reply:
I haven't seen such a scenario even though some of our zips are very big. 

Isn't the session extended when there is activity? 

Do you have reference to other popular web applications that do this?

Reply:
It is extended. Client does not know what is the TTL though. As a client: do you login again or do you reuse a session token after an hour, five hours, twelve hours of inactivity? Think about session as a temporary token: if it is temporary, it is import to tell the client how long is it valid.

Reply:
I see your point.

The client has the following choices that I can think of:
a. always login first
b. re-login if it gets an authentication error
Is it not enough?

Plenty of web applications use session tokens. I am curious how many of them use the patten you suggested.

Reply:
a. the benefit of using sessions is that the authentication is done when a session is not available/expires; what would be the benefit of having sessions when clients would authenticate before every action?
b. this leads to waste; from my experience, in many cases upload is the only action clients are performing via the API; on a slow VPN connections, this creates an unnecessary long round-trip

Reply:
> what would be the benefit of having sessions when clients would authenticate before every action?

Not every action.
A session allows multiple actions to be performed without authentication for each one. But not long enough to reduce security risk. 
Did I miss something?

Reply:
Currently, the open-source Hadoop Plugin which has an azkaban client is performing **b. re-login if it gets an authentication error** in most of the Gradle tasks it supports. 
@mtrna As we've discussed in https://github.com/linkedin/linkedin-gradle-plugin-for-apache-hadoop/pull/199 the Hadoop Plugin checks if authentication is valid by making an API call to createProject and then to upload. 
Ref: https://github.com/linkedin/linkedin-gradle-plugin-for-apache-hadoop/blob/master/hadoop-plugin/src/main/groovy/com/linkedin/gradle/azkaban/AzkabanUploadTask.groovy#L66

Let us know if you have a different use-case which shall help clients with information about TTL.

Reply:
@HappyRay : I was reacting to "always login first": it gives some space for interpretation.

In summary: what I find is missing on the client side is the information about whether the session token is still valid or not. Apparently we disagree on whether this information should be available to the clients of Azkaban. I am closing the PR then.

Reply:
Got it.

Actually, I like the changes you made except the line "ret.put("session.ttl", String.valueOf(cache.getEffectiveSessionTimeToLive()));" I understand it may have some value for some clients. But I am not convinced that the benefit is worth the lifetime cost of maintaining this additional contract. 

If you create a new PR with the line referenced above removed, I would be happy to review again and merge it. 

It would be even better if you can move the key name session.time.to.live to the common class that defines configuration keys. We would like to gradually migrate the existing constants there as the related code is changed.  

Thanks

Reply:
Hi @HappyRay ! Sure, the follow up PR with doc/config update only: https://github.com/azkaban/azkaban/pull/1626